src/jit/lsraxarch.cpp

```diff
@@ -2497,32 +2497,71 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree, TreeNodeInfo* info)
 void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, TreeNodeInfo* info)
 {
     NamedIntrinsic intrinsicID = intrinsicTree->gtHWIntrinsicId;
-    InstructionSet isa         = compiler->isaOfHWIntrinsic(intrinsicID);
+    InstructionSet isa         = Compiler::isaOfHWIntrinsic(intrinsicID);
     if (isa == InstructionSet_AVX || isa == InstructionSet_AVX2)
     {
         SetContainsAVXFlags(true, 32);
     }
-    info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp1);
-    if (intrinsicTree->gtGetOp2IfPresent() != nullptr)
+    GenTree* op1   = intrinsicTree->gtOp.gtOp1;
+    GenTree* op2   = intrinsicTree->gtOp.gtOp2;
+    info->srcCount = 0;
+
+    if (op1 != nullptr)
     {
-        info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp2);
+        if (op1->OperIsList())
+        {
+            for (GenTreeArgList* list = op1->AsArgList(); list != nullptr; list = list->Rest())
+            {
+                info->srcCount += GetOperandInfo(list->Current());
+            }
+        }
+        else
+        {
+            info->srcCount += GetOperandInfo(op1);
+        }
     }
 
-#ifdef _TARGET_X86_
-    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
+    if (op2 != nullptr)
     {
-        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS can be used as byte registers.
-        //
-        // TODO - currently we use the BaseType to bring the type of the second argument
-        // to the code generator. May encode the overload info in other way.
-        var_types srcType = intrinsicTree->gtSIMDBaseType;
-        if (varTypeIsByte(srcType))
+        info->srcCount += GetOperandInfo(op2);
+    }
+
+    switch (intrinsicID)
+    {
+        case NI_SSE41_BlendVariable:
         {
-            LocationInfoListNode* op2Info = useList.GetSecond(INDEBUG(intrinsicTree->gtGetOp2()));
-            op2Info->info.setSrcCandidates(this, RBM_BYTE_REGS);
+            if (!compiler->canUseVexEncoding())
+            {
+                // SSE4.1 blendv* hardcode the mask vector (op3) in XMM0
+                LocationInfoListNode* op3Info = useList.Begin()->Next()->Next();
+                op3Info->info.setSrcCandidates(this, RBM_XMM0);
+                op3Info->info.isDelayFree = true;
+                // so the dst register of blendv* cannot be XMM0
+                info->setDstCandidates(this, allSIMDRegs() & ~RBM_XMM0);
+                info->hasDelayFreeSrc = true;
```

Comment:
@mikedn Does here look correct to you?

This code works fine for common cases, but cannot work with BlendVariable chains. 
```csharp
var vf1 = Unsafe.Read<Vector128<float>>(ptr1);
var vf2 = Unsafe.Read<Vector128<float>>(ptr2);
var vf3 = Unsafe.Read<Vector128<float>>(ptr3);
var vf5 = Sse41.BlendVariable(vf1, vf3, vf2);
var vf6 = Sse41.BlendVariable(vf1, vf5, vf5);
var vf7 = Sse41.BlendVariable(vf1, vf2, vf6);
```
the above code generates
```asm
IN0053: 0001E1 movups   xmm0, xmmword ptr [rax]
IN0054: 0001E4 movups   xmm1, xmmword ptr [rdi+16]
IN0055: 0001E8 movaps   xmmword ptr [V05 rbp-A0H], xmm0
IN0056: 0001EF movaps   xmm1, xmm6
IN0057: 0001F2 blendvps  xmm1, xmm1
IN0058: 0001F7 movaps   xmm2, xmm6
IN0059: 0001FA blendvps  xmm2, xmm1
IN005a: 0001FF movaps   xmm1, xmmword ptr [V05 rbp-A0H]
IN005b: 000206 blendvps  xmm6, xmm1
```

Seems we still need "copy op3 to XMM0" for the corner cases.

Reply:
> Seems we still need "copy op3 to XMM0" for the corner cases.

Hmm, not sure if xmm0 needs to be copied but your example has an obvious problem with op1 and op2 now:
```
IN0056: 0001EF movaps   xmm1, xmm6
IN0057: 0001F2 blendvps  xmm1, xmm1
```
`dst` ended up in `xmm1` but `xmm1` contains `op2`. That needs to be "delay free" too.

This actually looks like a side effect of these not behaving like proper binary operators. If you look at the end of `TreeNodeInfoInit` you'll see that there's a bit of code that applies to binary RMW operators (e.g. `add eax, ebx`). But that code doesn't work well in the case of `blendvps` because op2 is null. So you'll have to do the right thing in your blend specific code - set both op2 and op3 to be delay free.

> Does here look correct to you?

You don't need `info->setDstCandidates(this, allSIMDRegs() & ~RBM_XMM0);`, you only need to the delay free stuff.

> The current S.N.Vector<T> code also uses this approach for SimdIntrinsicInitN.

Beware that there's a difference between `SimdIntrinsicInitN` and `blendvps`, at least when it comes to register allocation. It looks like `SimdIntrinsicInitN` uses an internal register to build the destination value. By the time it copies the value in the internal register to the dst register the source registers are no longer needed so it doesn't have this particular problem.

Reply:
Do we have tests to verify we emit in the above cases correct code?